### ISSUE TITLE
racket-mode's repl and coterm mode hate each other

### ISSUE THREAD

**Issue by zenspider** [greghendershott/racket-mode#665]
without coterm-mode:
```
; 
; Welcome to Racket v8.9 [cs].
; 
> 
5
blah.rkt>
```
with coterm-mode:
```
; 
  ; Welcome to Racket v8.9 [cs].
                                ; 
                                  > 
                                    5
                                     blah.rkt> 
```

set via customize to ensure it is global

I've tried experimenting with these as my racket-repl-hook:

```
(coterm-mode -1)                           ; too late? infected?
(setq comint-inhibit-carriage-motion nil)  ; no diff
(setq coterm--t-log-buffer "coterm-log")   ; looks FINE
```

<details>
<pre>
((alist-get 'racket-mode package-alist))
((emacs-version "28.2")
 (system-type darwin)
 (x-gtk-use-system-tooltips UNDEFINED)
 (major-mode racket-repl-mode)
 (racket--el-source-dir "/Users/ryan/.emacs.d/elpa/racket-mode-20230508.1742/")
 (racket--rkt-source-dir "/Users/ryan/.emacs.d/elpa/racket-mode-20230508.1742/racket/")
 (racket-program "~/Applications/dev/lisp/Racket/bin/racket")
 (racket-command-timeout 10)
 (racket-path-from-emacs-to-racket-function UNDEFINED)
 (racket-path-from-racket-to-emacs-function UNDEFINED)
 (racket-browse-url-function racket-browse-url-using-temporary-file)
 (racket-documentation-search-location "https://docs.racket-lang.org/search/index.html?q=%s")
 (racket-xp-after-change-refresh-delay 1)
 (racket-xp-mode-lighter
  (:eval
   (racket--xp-mode-lighter)))
 (racket-xp-highlight-unused-regexp "^[^_]")
 (racket-repl-buffer-name-function nil)
 (racket-submodules-to-run
  ((test)
   (main)))
 (racket-memory-limit 2048)
 (racket-error-context medium)
 (racket-repl-history-directory "~/.emacs.d/racket-mode/")
 (racket-history-filter-regexp "\\`\\s *\\'")
 (racket-images-inline t)
 (racket-imagemagick-props nil)
 (racket-images-keep-last 100)
 (racket-images-system-viewer "open")
 (racket-pretty-print t)
 (racket-use-repl-submit-predicate nil)
 (racket-pretty-print t)
 (racket-indent-curly-as-sequence t)
 (racket-indent-sequence-depth 0)
 (racket-pretty-lambda nil)
 (racket-smart-open-bracket-enable nil)
 (racket-module-forms "\\s(\\(?:module[*+]?\\|library\\)")
 (racket-logger-config
  ((cm-accomplice . warning)
   (GC . info)
   (module-prefetch . warning)
   (optimizer . info)
   (racket/contract . error)
   (racket-mode-debugger . info)
   (sequence-specialization . info)
   (* . fatal)))
 (racket-show-functions
  (racket-show-echo-area)))
(enabled-minor-modes
 (auto-composition-mode)
 (auto-compression-mode)
 (auto-dim-other-buffers-mode)
 (auto-encryption-mode)
 (auto-fill-mode)
 (auto-save-mode)
 (column-number-mode)
 (coterm-auto-char-lighter-mode)
 (coterm-auto-char-mode)
 (coterm-mode)
 (cyphejor-mode)
 (electric-indent-mode)
 (file-name-shadow-mode)
 (font-lock-mode)
 (gcmh-mode)
 (global-auto-revert-mode)
 (global-eldoc-mode)
 (global-font-lock-mode)
 (global-whitespace-mode)
 (keyfreq-autosave-mode)
 (keyfreq-mode)
 (line-number-mode)
 (mac-auto-operator-composition-mode)
 (mac-mouse-wheel-mode)
 (midnight-mode)
 (mode-line-bell-mode)
 (save-place-mode)
 (savehist-mode)
 (semantic-minor-modes-format)
 (shell-dirtrack-mode)
 (show-paren-mode)
 (temp-buffer-resize-mode)
 (transient-mark-mode)
 (which-key-mode)
 (windmove-mode)
 (window-number-meta-mode)
 (winner-mode))
(disabled-minor-modes
 (abbrev-mode)
 (auto-fill-function)
 (auto-revert-mode)
 (auto-revert-tail-mode)
 (auto-save-visited-mode)
 (blink-cursor-mode)
 (buffer-face-mode)
 (buffer-read-only)
 (button-mode)
 (cl-old-struct-compat-mode)
 (compilation-minor-mode)
 (compilation-shell-minor-mode)
 (completion-in-region-mode)
 (context-menu-mode)
 (coterm-char-mode)
 (coterm-scroll-snap-mode)
 (dash-fontify-mode)
 (defining-kbd-macro)
 (delete-selection-mode)
 (diff-auto-refine-mode)
 (diff-minor-mode)
 (dired-hide-details-mode)
 (dispwatch-mode)
 (eldoc-mode)
 (electric-layout-mode)
 (electric-quote-mode)
 (flycheck-mode)
 (global-dash-fontify-mode)
 (global-flycheck-mode)
 (global-prettify-symbols-mode)
 (global-semantic-highlight-edits-mode)
 (global-semantic-highlight-func-mode)
 (global-semantic-show-parser-state-mode)
 (global-semantic-show-unmatched-syntax-mode)
 (global-semantic-stickyfunc-mode)
 (global-visual-line-mode)
 (global-whitespace-newline-mode)
 (horizontal-scroll-bar-mode)
 (hs-minor-mode)
 (indent-tabs-mode)
 (isearch-mode)
 (jit-lock-debug-mode)
 (lock-file-mode)
 (mac-auto-ascii-mode)
 (mac-font-panel-mode)
 (menu-bar-mode)
 (mouse-wheel-mode)
 (next-error-follow-minor-mode)
 (overwrite-mode)
 (p4-mode)
 (paragraph-indent-minor-mode)
 (paredit-mode)
 (prettify-symbols-mode)
 (racket-smart-open-bracket-mode)
 (racket-xp-mode)
 (semantic-highlight-edits-mode)
 (semantic-highlight-func-mode)
 (semantic-mode)
 (semantic-show-parser-state-mode)
 (semantic-show-unmatched-syntax-mode)
 (semantic-stickyfunc-mode)
 (sh-electric-here-document-mode)
 (size-indication-mode)
 (sticky-buffer-mode)
 (tab-bar-history-mode)
 (tab-bar-mode)
 (text-scale-mode)
 (tool-bar-mode)
 (tooltip-mode)
 (unify-8859-on-decoding-mode)
 (unify-8859-on-encoding-mode)
 (url-handler-mode)
 (use-hard-newlines)
 (view-mode)
 (visible-mode)
 (visual-line-mode)
 (whitespace-mode)
 (whitespace-newline-mode)
 (window-divider-mode)
 (window-number-mode)
 (winnow-mode)
 (xref-etags-mode))
</pre>
</details>

**Comment by zenspider** [comment]
Culprit commit: https://repo.or.cz/emacs-coterm.git/blobdiff/24ebe50850df6eb98d1a70f34b70d67e940c62ac..ef994718487abe2004aaf38a63660c76c224cfb1:/coterm.el

Turns out that `coterm--init` adds some wrappers around the process-filter that run `coterm--t-emulate-terminal`, whether coterm is enabled or not. I think it should at least be checking for `coterm-mode` first, or better, those wrappers should be removed when coterm is disabled.

workaround, add this to your racket-repl-mode-hook:

```elisp
(remove-function (process-filter (get-buffer-process (current-buffer)))
                 #'coterm--t-emulate-terminal)
```

that was a serious pain in the ass to figure out.

I don't know how to file a bug w/ this repo

**Comment by greghendershott** [comment]
> I don't know how to file a bug w/ this repo

I took a quick glance at `coterm.el` and [around line 110 it says](https://repo.or.cz/emacs-coterm.git/blob/HEAD:/coterm.el#l110):

```elisp
;;;; Bugs, suggestions and patches can be sent to
;;
;;    bugs-doseganje (at) groups.io
;;
;; and can be viewed at https://groups.io/g/bugs-doseganje/topics.  As this
;; package is stored in GNU ELPA, non-trivial patches require copyright
;; assignment to the FSF, see info node "(emacs) Copyright Assignment".
```

Without understanding the problem too deeply, my first idea is that `coterm` should limit automatically enabling itself based on a new customization variable that is a whitelist or blacklist of other modes.

**Comment by greghendershott** [comment]
@zenspider Did you want to email a bug to them?

Although I could, because I don't use `coterm` if they have any follow-up question about e.g. config/usage I'd be "uh, well...". (But happy to discuss details about its interop with `racket-repl-mode`.)

TL;DR: I'm not sure whether there's more that I could or should do with this; wondering whether to keep the issue open or close it.

**Comment by jakanakaevangeli** [comment]
I recieved a personal mail from zenspider. Thanks and sorry for late reply.

I should probably make coterm--init into a local minor mode so that the user can indeed turn it off in these cases.
I hope I'll be able to do it soon, but I'm still waiting for my FSF copyright papers.

Feel free to close this issue as I believe it relates to coterm rather than racket-mode.

**Comment by greghendershott** [comment]
Hi -- thank you for following up!

I'll go ahead and close this. But if there's anything I can do to help please don't hesitate to let me know.

**Comment by zenspider** [comment]
Yeah, I'll do it

**Comment by zenspider** [comment]
filed here: https://groups.io/g/bugs-doseganje/topic/racket_mode_and_coterm_hate/101401544

**Comment by greghendershott** [comment]
FWIW I think this might be moot now (for the last several weeks) that `racket-repl-mode` is no longer using a process or `comint-mode` for user program I/O.
